### PR TITLE
be:macos: Add simple backend for macOS

### PR DESCRIPTION
**Pull request by Baekalfen**
New backend which supports macOS mostly using existing POSIX functionality. Although with the addition to  use the NVMeSMARTLibExternal.h provided by Xcode to query some information from the target disk. This does not enable arbitrary NVMe commands to be sent to a disk, but rather support for Get Log Page and Identify commands.

### CODE REVIEW

**Comment by Baekalfen** [comment]
Blocked by #55

**Comment by safl** [comment]
Regarding the first commit. The issue you are experiencing is not a general ``aarch64`` issue.
Disabling the SPDK backend on all ``aarch`` is therefore not a good solution.
See here: https://github.com/OpenMPDK/xNVMe/pull/66
So, either:
1) Find the root cause of the issue and make a patch for SPDK, we can carry this patch:
  - ``subprojects/packagefiles/spdk/patches/`` until we get the patch upstreamed in SPDK
2) Alter the commit to do a work-around specific to where you are seeing the issue.

**Comment by safl** [comment]
Regarding the second commit ``be:psync,macos: unify pread/pwrite across Unix-like platforms``. Thanks for taking a stab at this, it is an area that would be really great to improve for re-use. However, I am not certain this commit is the way to go. There are some things that need a bit more investigation when it comes the platform-specific behavior and error-handling on even the simplest things like ``pread()``/``pwrite()``.
Also, on MacOS, it looks like some special handling is needed for fsync():
* https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/fsync.2.html#//apple_ref/doc/man/2/fsync
* https://twitter.com/marcan42/status/1494213862970707969

So, instead of the refactor, how about just doing this:
* export the POSIX ``cmd_io()`` implementation and re-use it in ``lib/xnvme_be_macos_sync.c``
* add cmd_iov() in ``lib/xnvme_be_macos_sync.c``

Right now, I would prefer not to alter the behavior of all of POSIX, Linux, and FreeBSD for the sake of the MacOS implementation. I would much rather like that the MacOS backend is added and then we can investigate a proper re-usable ``common``/``base`` implementation. Taking other things into account such as the ``common/base`` backend state and other things that are commonly needed/re-used.

**TLDR**; to unblock the addition of the MacOS addition, then just add the MacOS implementation and re-use the ``cmd_io()`` from the POSIX backend.

**Comment by Baekalfen** [comment]
> Regarding the first commit `build:meson: exclude SPDK from building on aarch64 systems`. The issue you are experiencing is not a general `aarch64` issue. Disabling the SPDK backend on all `aarch` is therefore not a good solution. See here: #66 So, either:
> 
> 1. Find the root cause of the issue and make a patch for SPDK, we can carry this patch:
> 
> * `subprojects/packagefiles/spdk/patches/` until we get the patch upstreamed in SPDK
> 
> 2. Alter the commit to do a work-around specific to where you are seeing the issue.

I'll drop it from here. It was just part of the branch before you made the comments in #66

**Comment by Baekalfen** [comment]
> Regarding the second commit `be:psync,macos: unify pread/pwrite across Unix-like platforms`. Thanks for taking a stab at this, it is an area that would be really great to improve for re-use. However, I am not certain this commit is the way to go. There are some things that need a bit more investigation when it comes the platform-specific behavior and error-handling on even the simplest things like `pread()`/`pwrite()`. Also, on MacOS, it looks like some special handling is needed for fsync():
> 
> * https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/fsync.2.html#//apple_ref/doc/man/2/fsync
> * https://twitter.com/marcan42/status/1494213862970707969
> 
> So, instead of the refactor, how about just doing this:
> 
> * export the POSIX `cmd_io()` implementation and re-use it in `lib/xnvme_be_macos_sync.c`
> * add cmd_iov() in `lib/xnvme_be_macos_sync.c`
> 
> Right now, I would prefer not to alter the behavior of all of POSIX, Linux, and FreeBSD for the sake of the MacOS implementation. I would much rather like that the MacOS backend is added and then we can investigate a proper re-usable `common`/`base` implementation. Taking other things into account such as the `common/base` backend state and other things that are commonly needed/re-used.
> 
> **TLDR**; to unblock the addition of the MacOS addition, then just add the MacOS implementation and re-use the `cmd_io()` from the POSIX backend.

Maybe I should have pointed it out more clearly. That commit is actually from #55

**Comment by Baekalfen** [comment]
@safl If you have no more comments, then I consider this PR ready for merge.

File: toolbox/cijoe-pkg-xnvme/testplans/xnvme-base-macos.plan [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  * admin-commands idfy-ns, idfy-ctrlr, get-log, get/set feature
+  * queue-init/teardown of io_uring/libaio/posix
+  Requires
+  * XNVME_URI points to an NVMe namespace e.g. /dev/nvme0n1
+evars:
+  NVME_NSTYPE: "lblk"
+  DEV_TYPE: "nvme"
+  XNVME_BE: "macos"
+  XNVME_ADMIN: "macos"
+  XNVME_URI: "/dev/disk4"
```
Comment (safl) [inline toolbox/cijoe-pkg-xnvme/testplans/xnvme-base-macos.plan:16]: The ``XNVME_URI`` be removed from here and instead adjust the construction of ``XNVME_URI`` in ``toolsbox/cijoe-pkg-xnvme/envs/refenv-xnvme.sh`` and ``toolsbox/cijoe-pkg-xnvme/envs/refenv-xnvme.sh`.

File: toolbox/cijoe-pkg-xnvme/testcases/examples-xnvme_hello.sh [hunk 2]
```diff
@@ -1,4 +1,4 @@
-#!/bin/bash
```
Comment (safl) [inline toolbox/cijoe-pkg-xnvme/testcases/examples-xnvme_hello.sh:1]: The commit title should be something along the lines of: ``toolbox:cijoe: change shebang to '/usr/bin/env bash'``

And the body should describe why this is needed.

**Review by safl — CHANGES_REQUESTED** [review]
See in-line comments, adding to those:
* Could the introduction of ``additional_link_args`` / ``link_args`` in meson be handled in a manner similar to ``dependency('threads')``?

**Review by safl — APPROVED** [review]

**Comment by safl** [comment]
Thanks!
